ADR/0033-upstreaming-integration-service.md

```diff
@@ -0,0 +1,270 @@
+carries this `CONTROLLABEL` label the **control** TestSubject for the `LABELNAME` label.
+
+[integration-service] should grow a mutating webhook that *removes* the `CONTROLLABEL` label from
+the last control [TestSubject] for the `LABELNAME` label, when the `CONTROLLABEL` label is applied
+to a new [TestSubject] bearing the `LABELNAME` label.
+
+The **control** TestSubject is used as the *basis* for new TestSubject construction. See
+"TestSubject Construction as optional behavior", below.
+
+[integration-service] should promote a [TestSubject] under test to become the new **control**
+TestSubject after it passes all of the required IntegrationTestScenarios with the matching
+`LABELNAME` label.
+
+### TestSubject Construction as optional behavior
+
+**Test Initiation** has a clear control plane: the [IntegrationTestScenario] combined with the
+[TestSubject]. Meaning, when the user creates an [IntegrationTestScenario] and a [TestSubject], and
+the [selector] of the [IntegrationTestScenario] matches the attributes of [TestSubject], then together
+they represent intent for the subject to be tested. The user uses these resources to control what
+the system does.
+
+Today, control of the **Snapshot Construction** process is mixed up in the [Application] and [Component]
+model, and is partly just implied. We *always* construct [Snapshots].
+
+**TestSubject Construction** (called Snapshot Construction in our system today) should have a clear
+control plane resource: the [TestSubjectConstructor]. A hypothetical schema for the resource
+follows:
+
+```yaml
+version: appstudio.redhat.com/v1alpha1
+kind: TestSubjectConstructor
+metadata:
+  name: my-application-test-subject-constructor
+spec:
+  selector:
+    fields:
+      match:
+        kind: PipelineRun
+        version: tekton.dev/v1
+    labels:
+      match:
+        "appstudio.openshift.io/application": my-application
+        "pipelinesascode.tekton.dev/state": completed
+  extractor:
+    image_url: '.status.results.[] | select(.name == "IMAGE_URL").value'
+    image_digest: '.status.results.[] | select(.name == "IMAGE_DIGEST").value'
+    name: '.metadata.labels."appstudio.openshift.io/component"'
+    source:
+      git:
+        revision: '.metadata.labels."pipelinesascode.tekton.dev/sha"'
+        url: '.metadata.labels."pipelinesascode.tekton.dev/repo-url"'
+```
+
+Don't take the fields literally. The details will likely change after the ADR has settled. It is
+here for illustration purposes.
+
+The `selector` field will be used to identify resources that should trigger the construction of new
+[TestSubjects]. When created, the *basis* for the new test subject should always be taken from the
+**control** [TestSubject]. The values extracted from the triggering resource by way of the
+expressions on the `extractor` resource should be applied to that basis in the construction of a new
+[TestSubject]. If there is no **control** [TestSubject], then the new TestSubject should be created
```

Comment:
If we are loosening the dependency on Applications and Components, how do we know how to convert from a PipelineRun to a resource in the TestSubject?

Reply:
I think so. This is the `extractors` field in the hypothetical spec of the TestSubjectConstructor.

Reply:
Right. This question is less relevant with the other comments that I have. I made it earlier in the review and didn't refine it.